packages/remix-react/components.tsx

```diff
@@ -1022,12 +1022,22 @@ export function useFormAction(
   // https://github.com/remix-run/remix/issues/927
   let location = useLocation();
   let { search, hash } = resolvedPath;
+  let isIndexRoute = id.endsWith("/index");
+
   if (action == null) {
-    search = location.search;
+    // When grabbing search params from the URL, remove the automatically
+    // inserted index param so we match the useResolvedPath search behavior
+    // which would not include ?index
+    if (isIndexRoute) {
+      let params = new URLSearchParams(location.search);
+      params.delete("index");
+      search = params.toString() ? `?${params.toString()}` : "";
+    } else {
+      search = location.search;
+    }
     hash = location.hash;
   }
 
-  let isIndexRoute = id.endsWith("/index");
   if ((action == null || action === ".") && isIndexRoute) {
     search = search ? search.replace(/^\?/, "?index&") : "?index";
```

Comment:
This logic relies on search _not_ having the index param in it, which is how `useResolvedPath` returns it.  Now that we pull directly from `useLocation`, it _does_ have it so this logic appends a duplicate one.  I chose to remove it from the new code path above and leave this untouched, but we could also likely just use separate logic for the `action == null` and `action === "."` cases here if that would be clearer?